### ISSUE TITLE
Support a way to extend the electron process

### ISSUE THREAD

**Issue by matthewmueller** [segmentio/nightmare#391]
initial thoughts:

- support adding requires that pass in the window

```js
electron.plugins.forEach(function (plugin) {
  require(plugin)(win)
})
```

This would happen after the browser window has been initialized

**Comment by rosshinkley** [comment]
It might also make sense to pass `parent`.  That would grant implementing custom actions in Nightmare with custom downstream actions in the Electron process.  Thoughts?

---

Pondering out loud: is there a reason to prefer using `require`?  If so, why does `action` not offer a `require` option?  Would the plugin paths be passed in with options, or added later?

**Comment by bjrmatos** [comment]
It would make more sense to implement #393 and let nigthmare focus on its main job

**Comment by bjrmatos** [comment]
inevitably, people will start to ask more and more features outside nightmare's job, it will be more easy to let nigthmare run as a electron module and developers can do what they want with the electron process

**Comment by matthewmueller** [comment]
@rosshinkley i don't see a way you'd be able to pass `parent`, it's transferring over IPC and parent isn't serializable. 

The reason for doing the require is that we could basically tell the electron process to require a file on it's own. I don't like the solution all that much, but I haven't come up with a better one.

**Comment by rosshinkley** [comment]
@matthewmueller did I misunderstand?  I thought the intent was passing paths to require over IPC (or possibly, as command-line arguments akin to `paths`) and requiring them in the electron process, not vice-versa.  No IPC serialization required.  Parent should then be passable.

I'll need to take another pass at this when I'm fresh and have coffee.  I'll try to carve out some time.

**Comment by matthewmueller** [comment]
>  I thought the intent was passing paths to require over IPC (or possibly, as command-line arguments akin to paths) and requiring them in the electron process, not vice-versa. No IPC serialization required. Parent should then be passable.

Ahh, sorry I thought you meant `parent` as in the parent process instance. Yep the original intent is passing paths to files and requiring them in the electron process, though I don't necessarily like this approach, so if you think of something better, I'm all ears. Can you clarify what you mean by `parent` because I'm still not fully sure?

> I'll need to take another pass at this when I'm fresh and have coffee. I'll try to carve out some time.

Haha yah me too, answering issues at this hour is ill-advised :-P

**Comment by rosshinkley** [comment]
@matthewmueller I should have been more specific, and the naming conventions are also probably not helping: in `runner`, `parent` is defined as the IPC process.

Re ideas for implementation (I still need to carve out time to experiment): how do you want them?  Should I make a branch on my fork and link to it, set up a PR with "proposal" in the header, or something else?  What would you prefer?

---

@bjrmatos I'm on board for having Nightmare be supported inside of Electron (#393), but I don't think you'd want to discard Nightmare as a node module.  (I would argue it's one of the bigger problems with using Phantom/Casper et al.  Usage from node is certainly possible, but a bigger pain.)  I agree that extensibility is currently an issue, but not an insurmountable one.

**Comment by matthewmueller** [comment]
@rosshinkley yah, terrible names haha. yah, I'd like to see what you have in mind. PR or commenting in this issue seem reasonable. Would be nice to see some example code, like how a dev would extend it.

**Comment by bjrmatos** [comment]
@rosshinkley my intention never was to discard using `nigthmare` as a node module :smile: however for tasks in scale the current state of `nigthmare` is not very efficient, i'm just trying to support a way to run `nigthmare` that let me scale in other ways

**Comment by rosshinkley** [comment]
Implemented in #425, merged in 2.3.0.  Closing.
